### PR TITLE
KAFKA-3631: Fix Struct.toString for nullable arrayOf

### PR DESCRIPTION
**Pull request by granthenke**

### CODE REVIEW

File: clients/src/test/java/org/apache/kafka/common/protocol/types/ProtocolSerializationTest.java [hunk 1]
```diff
@@ -231,6 +231,17 @@ public void testReadNegativeBytesSize() {
...
 
+    @Test
+    public void testToString() {
+        try {
+            String stuctStr = this.struct.toString();
+            assertFalse("Stuct string should not be null.", stuctStr == null);
+            assertFalse("Stuct string should not be empty.", stuctStr.isEmpty());
+        } catch (Throwable e) {
+            fail("Struct.toString throws an exception");
+        }
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/common/protocol/types/ProtocolSerializationTest.java]: Is it really worth having this catch here? I think it's best to just let the exception propagate. Any method under test can throw an unknown exception after all.

File: clients/src/main/java/org/apache/kafka/common/protocol/types/Struct.java [hunk 2]
```diff
@@ -292,13 +292,15 @@ public String toString() {
             b.append('=');
             if (f.type() instanceof ArrayOf) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/protocol/types/Struct.java]: Would it not be better to add a `&& this.values[i] != null` here?

File: clients/src/test/java/org/apache/kafka/common/protocol/types/ProtocolSerializationTest.java (hunk 1 above)
Comment (granthenke) [inline clients/src/test/java/org/apache/kafka/common/protocol/types/ProtocolSerializationTest.java]: I can remove it.

File: clients/src/test/java/org/apache/kafka/common/protocol/types/ProtocolSerializationTest.java [hunk 3]
```diff
@@ -231,6 +231,13 @@ public void testReadNegativeBytesSize() {
         }
     }
 
+    @Test
+    public void testToString() {
+        String stuctStr = this.struct.toString();
+        assertFalse("Stuct string should not be null.", stuctStr == null);
```
Comment (ijuma) [inline clients/src/test/java/org/apache/kafka/common/protocol/types/ProtocolSerializationTest.java]: `assertNotNull`?

File: clients/src/main/java/org/apache/kafka/common/protocol/types/Struct.java (hunk 2 above)
Comment (granthenke) [inline clients/src/main/java/org/apache/kafka/common/protocol/types/Struct.java]: I guess that would drop into the else statement, which works but I am not sure is "better" of different enough to matter. 

Why would you prefer it in the first if check?

Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/protocol/types/Struct.java]: Your approach doesn't append the `null` string or am I missing something?

Comment (granthenke) [inline clients/src/main/java/org/apache/kafka/common/protocol/types/Struct.java]: Ah sure. Will fix.

File: clients/src/test/java/org/apache/kafka/common/protocol/types/ProtocolSerializationTest.java (hunk 3 above)
Comment (granthenke) [inline clients/src/test/java/org/apache/kafka/common/protocol/types/ProtocolSerializationTest.java]: Sure

**Comment by SinghAsDev** [comment]
LGTM!

**Comment by ijuma** [comment]
LGTM

**Pull request closed without merging**